### DISCUSSION TITLE
Failed for 3D non-Cartesian trajectory

### DISCUSSION THREAD

**Discussion by Michaelsqj** [mmuckley/torchkbnufft#56] [Q&A]
Hi there,

I tried to use the torchkbnufft for 3D non-Cartesian trajectory for simulation. I used a 1x1x64x64x64 image and 3D radial trajectory as input to kbNufft for forward NUFFT and then use the acquired kdata for adjoint NUFFT, the output image doesn't look right. I'm wondering if torchkbnufft can be used for 3D non-Cartesian reconstruction? If so, are there any examples for 3D non-Cartesian reconstruction using this package?

Yours,
Qijia

**Comment by mmuckley** [comment]
Converting to a discussion as this does not seem to document a bug.

Hello @Michaelsqj, what are the ranges for your trajectory? As the documentation states, your trajectory needs to be in terms of radians/voxel. Also, you should ensure that you have enough samples in your trajectory to reconstruct without aliasing.

**Reply by Michaelsqj** [reply]
Hi, @mmuckley , Thank you so much for your reply! Sorry, it was completely my fault. I'm not quite familiar with the package and made mistakes at converting to radians/voxels, and I just succeeded the reconstruction. There are several additional questions I might have, and I'd really appreciate it if you could provide any thoughts!

1. I noticed that your torchkbNufft is inspired by Professor Jeff Fessler's irt package. I've been using irt package, and I'm very interested in moving to torchkbNufft for faster reconstruction. I'm wondering if you have compared the results of forward/ adjoint NUFFT with the irt package?

2. Additionally, does torchkbnufft require much GPU memory? I'm currently using GPU with 11GB memory, I'm wondering if you have rough idea about whether it's sufficient to handle most 3D non-Cartesian reconstruction case?

3. When GPU is not available, does torchkbnufft package also have great advantage in NUFFT speed comparing to the irt package?

Cannot be more grateful for your help!

**Reply by mmuckley** [reply]
Hello @Michaelsqj, happy to hear you figured it out!

1. I manually verified accuracy within numerical precision of Prof. Fessler's package prior to the original upload of `torchkbnufft`. Now, there is a [unit test](https://github.com/mmuckley/torchkbnufft/blob/b842f1303d4e08edd2e9aeb139c1375b3a081404/tests/test_interp.py#L16-L31) for the interpolator with test data from an earlier version of the package to ensure we never have an accuracy regression. All other accuracy tests in the package indirectly depend on this one.
2. I have tried to make the implementation efficient and use as few array allocations as possible, but of course memory usage will depend on your trajectory. I think applying NUFFTs with an 11 GB GPU should not be too bad, but note that doing 3D imaging work with neural networks is extremely memory-intensive.
3. Without a GPU, `torchkbnufft` may be faster or slower than CPU-based packages. There are some threading features that might help you with dynamic imaging, for example. See [Performance Tips](https://torchkbnufft.readthedocs.io/en/stable/performance.html) for how to use the package's features. The general summary is that: for cases where parallelization is straightforward across time or coils, you might be able to make `torchkbnufft` faster. For very large trajectories (like 3D imaging possibly), `torchkbnufft` tends to be slower.
